lightning/src/ln/wire.rs

```diff
@@ -19,13 +19,14 @@
 //! [BOLT #1]: https://github.com/lightningnetwork/lightning-rfc/blob/master/01-messaging.md
 
 use ln::msgs;
+use ln::peer_handler::CustomMessageReader;
 use util::ser::{Readable, Writeable, Writer};
 
 /// A Lightning message returned by [`read()`] when decoding bytes received over the wire. Each
 /// variant contains a message from [`msgs`] or otherwise the message type if unknown.
 #[allow(missing_docs)]
 #[derive(Debug)]
-pub enum Message {
+pub enum Message<T> where T: std::fmt::Debug {
```

Comment:
Why store a `T` in `Custom`? Can't we get away with just making `Custom` store a `Vec<u8>` instead, avoiding a ton of generic typing for a similar outcome?

Reply:
> reduces the amount of code that needs to change in peer handler

I'm pretty dubious of this - there's a lot of new generic types flying everywhere and we already need to call a new method in the decode pipeline anyway. Maybe more importantly, the new API isn't trivially composable. Imagine I want to have a `DLCMessageHandler` *and* a `MattsToSMessageHandler` attached at once - I can't just match the message type and pass the message onwards to the next handler, I have to create a whole new enum that maps to multiple different potential message enums, which blows up the complexity substantially.

Reply:
> I'm pretty dubious of this - there's a lot of new generic types flying everywhere and we already need to call a new method in the decode pipeline anyway.

The proposed changes to `peer_handler` seemed quite extensive comparatively: https://github.com/p2pderivatives/rust-lightning/pull/3.

The approach in this PR is an alternative that I thought was much simpler. Additionally, separating parsing from handling seemed a cleaner separation of abstractions similar to `wire` and `ChannelMessageHandler`.

Happy to consider other alternatives.

> Maybe more importantly, the new API isn't trivially composable. Imagine I want to have a `DLCMessageHandler` _and_ a `MattsToSMessageHandler` attached at once - I can't just match the message type and pass the message onwards to the next handler, I have to create a whole new enum that maps to multiple different potential message enums, which blows up the complexity substantially.

We should be able to implement the trait on a 2-tuple like we do for `chain::Listen`. It would require defining an enum with two variants to use for the return value. The implementation of `read` would cascade as needed. And by nesting tuples it could be arbitrarily composable.

I suppose we could also use a trait object for the return type if you want to reduce the number of generic types. Figured we wanted to avoid `dyn` however.

Reply:
W.r.t to composability I did have in mind to push the conversation behind the `UMH` interface. One implementation could be a proxy chaining processing calls on a vec of dynamic objects ?